hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ConditionalWriteLockProvider.java

```diff
@@ -0,0 +1,581 @@
+  @GuardedBy("this")
+  private StorageLockFile currentLockObj = null;
+  // Ensures we do not try to lock after being closed.
+  @GuardedBy("this")
+  private boolean isClosed = false;
+
+  private synchronized void setLock(StorageLockFile lockObj) {
+    if (lockObj != null && !Objects.equals(lockObj.getOwner(), this.ownerId)) {
+      throw new HoodieLockException("Owners do not match! Current LP owner: " + this.ownerId + " lock path: "
+          + this.lockFilePath + " owner: " + lockObj.getOwner());
+    }
+    this.currentLockObj = lockObj;
+  }
+
+  /**
+   * Default constructor for ConditionalWriteLockProvider, required by LockManager
+   * to instantiate it using reflection.
+   * 
+   * @param lockConfiguration The lock configuration, should be transformable into
+   *                          ConditionalWriteLockConfig
+   * @param conf              Storage config, ignored.
+   */
+  public ConditionalWriteLockProvider(final LockConfiguration lockConfiguration, final StorageConfiguration<?> conf) {
+    ConditionalWriteLockConfig config = new ConditionalWriteLockConfig.Builder()
+        .fromProperties(lockConfiguration.getConfig()).build();
+    heartbeatIntervalMs = config.getHeartbeatPollMs();
+    lockValidityMs = config.getLockValidityTimeoutMs();
+
+    // Determine if the provided locks location is relative.
+    String configuredLocksLocation = config.getLocksLocation();
+
+    // If using base path, recalculate the locks location as .hoodie/.locks
+    String locksLocation = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? String.format("%s%s%s", config.getHudiTableBasePath(), StoragePath.SEPARATOR, LOCKS_FOLDER_NAME)
+        : configuredLocksLocation;
+
+    URI uri = parseURI(locksLocation);
+    bucketName = uri.getHost(); // For most schemes, the bucket/container is the host.
+    String folderName = uri.getPath(); // Path after the bucket/container.
+
+    String fileName = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? DEFAULT_TABLE_LOCK_FILE_NAME
+        : slugifyLockFolderFromBasePath(config.getHudiTableBasePath());
+
+    lockFilePath = buildLockObjectPath(folderName, fileName);
+    ownerId = UUID.randomUUID().toString();
+    this.logger = DEFAULT_LOGGER;
+    this.heartbeatManager = new LockProviderHeartbeatManager(
+        ownerId,
+        heartbeatIntervalMs,
+        this::renewLock);
+
+    try {
+      this.lockService = (StorageLock) ReflectionUtils.loadClass(
+          getLockServiceClassName(uri.getScheme()),
+          new Class<?>[] { String.class, String.class, String.class, Properties.class },
+          new Object[] { ownerId, bucketName, lockFilePath, lockConfiguration.getConfig() });
+    } catch (Throwable e) {
+      throw new HoodieLockException("Failed to load and initialize StorageLock", e);
+    }
+
+    logger.info("Instantiated new Conditional Write LP, owner: {}, lockfilePath: {}", ownerId, lockFilePath);
```

Comment:
Avoid abbreviation in the logs shown to the user which can be confusing
```suggestion
    logger.info("Instantiated new storage-based lock provider, owner: {}, lockfilePath: {}", ownerId, lockFilePath);
```